### PR TITLE
Add codegen test for issue #73827

### LINKED ISSUE #73827: Missed optimization to elide bounds check
Context: this bug was reduced from indexing into a nested array, but this shows the same problem.

```rust
pub fn get(array: &[u8; 8], x: usize, y: usize) -> u8 {
    if x > 7 || y > 7 {
        0
    } else {
        array[y]
    }
}
```
([Playground link](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=eb99add1d3d6d8341bbb85feb75a5781))

Rust doesn't elide the bounds check on `array[y]`, even though it's impossible for `y` to be out of range due to the if condition. Rust does elide the bounds check when the if condition is just `if y > 7`, however.

Here is the generated ASM for the above code:
```asm
playground::get:
	pushq	%rax
	orq	%rdx, %rsi
	cmpq	$7, %rsi
	jbe	.LBB0_2
	xorl	%eax, %eax
	popq	%rcx
	retq

.LBB0_2:
	cmpq	$7, %rdx
	ja	.LBB0_5
	movb	(%rdi,%rdx), %al
	popq	%rcx
	retq

.LBB0_5:
	leaq	.Lanon.357673f8919c00a2ec2e627b7663c19f.1(%rip), %rax
	movl	$8, %esi
	movq	%rdx, %rdi
	movq	%rax, %rdx
	callq	*core::panicking::panic_bounds_check@GOTPCREL(%rip)
	ud2

.Lanon.357673f8919c00a2ec2e627b7663c19f.0:
	.ascii	"src/lib.rs"

.Lanon.357673f8919c00a2ec2e627b7663c19f.1:
	.quad	.Lanon.357673f8919c00a2ec2e627b7663c19f.0
	.asciz	"\n\000\000\000\000\000\000\000\005\000\000\000\t\000\000"
```

and here is the generated ASM when we change the if to just be `if y > 7`
```asm
playground::get:
	cmpq	$7, %rdx
	jbe	.LBB0_2
	xorl	%eax, %eax
	retq

.LBB0_2:
	movb	(%rdi,%rdx), %al
	retq
```

### Meta
This occurs in both the latest versions of stable (1.44.1) and nightly (1.46.0-nightly (2020-06-26 7750c3d46bc19784adb1))

### PR DESCRIPTION
**Pull request by bugadani**
Closes #73827

### CODE REVIEW

File: src/test/codegen/issue-73827-bounds-check-index-in-subexpr.rs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+#![crate_type = "lib"]
+
+// CHECK-LABEL: @get
+#[no_mangle]
+pub fn get(array: &[u8; 8], x: usize, y: usize) -> u8 {
+    if x > 7 || y > 7 {
+        0
+    } else {
+        // CHECK-NOT: panic_bounds_check
```
Comment (lzutao) [inline src/test/codegen/issue-73827-bounds-check-index-in-subexpr.rs:15]: I'd prefer to move this line up and move it below the CHECK-LABEL.

Comment (bugadani) [reply src/test/codegen/issue-73827-bounds-check-index-in-subexpr.rs:15]: I don't want to write this as a dismissal, but I personally prefer having them where I expect the instruction to (not) be.

Comment (lzutao) [reply src/test/codegen/issue-73827-bounds-check-index-in-subexpr.rs:15]: That is fine. I just prefer to do whatever <https://llvm.org/docs/CommandGuide/FileCheck.html> does.

Comment (bugadani) [reply src/test/codegen/issue-73827-bounds-check-index-in-subexpr.rs:15]: I guess when you have two languages that are close enough, mixing them could seem messy. But with Rust I feel marking the problematic point can have a tiny bit of usefulness :)

**Comment by Mark-Simulacrum** [comment]
r? @nikic 

I think this is right though negative testing is always a bit concerning (if we were to rename panic_bounds_check it seems like this test would continue to pass, despite being wrong). Have we verified that it fails on older rustc compilers (that didn't have LLVM 11)?

**Comment by nikic** [comment]
@bors r+ rollup

Yeah, negative testing can fail if the function gets renamed, but generating full positive check lines couples us too tightly to LLVM. We already have plenty of other tests that use CHECK-NOT: panic_bounds_check, so I think it's fine to go with this.

**Pull request merged**